### PR TITLE
add ExpData __repr__ 

### PR DESCRIPTION
**Pull request by FFroehlich**

### CODE REVIEW

File: swig/edata.i [hunk 1]
```diff
@@ -8,5 +8,51 @@ using namespace amici;
...
+        for it in range(self.nt())
+        for iy in range(self.nytrue())
+    )
+    n_data_z = sum(
+        self.isSetObservedData(ie, iz)
+        for ie in range(self.nmaxevent())
+        for iz in range(self.nztrue())
+    )
+    n_sigma_z = sum(
+        self.isSetObservedDataStdDev(ie, iz)
```
Comment (dweindl) [inline swig/edata.i:30]:
```suggestion
        self.isSetObservedEventsStdDev(ie, iz)
```

File: swig/edata.i [hunk 2]
```diff
@@ -8,5 +8,51 @@ using namespace amici;
...
+        for it in range(self.nt()) for
+        iy in range(self.nytrue())
+    )
+    n_sigma_y = sum(
+        self.isSetObservedDataStdDev(it, iy)
+        for it in range(self.nt())
+        for iy in range(self.nytrue())
+    )
+    n_data_z = sum(
+        self.isSetObservedData(ie, iz)
```
Comment (dweindl) [inline swig/edata.i:25]:
```suggestion
        self.isSetObservedEvents(ie, iz)
```

File: swig/edata.i [hunk 3]
```diff
@@ -8,5 +8,51 @@ using namespace amici;
...
+        for ie in range(self.nmaxevent())
+        for iz in range(self.nztrue())
+    )
+    return "\n".join([
+        self.this.__repr__()[:-1],
+        f"  {self.nt()}x{self.nytrue()} time-resolved datapoints",
+        f"    ({n_data_y}/{self.nt()*self.nytrue()} measurements & {n_sigma_y}/{self.nt()*self.nytrue()} sigmas set)",
+        f"  {self.nmaxevent()}x{self.nztrue()} event-resolved datapoints",
+        f"    ({n_data_z}/{self.nmaxevent()*self.nztrue()} measurements & {n_sigma_z}/{self.nmaxevent()*self.nztrue()} sigmas set)",
+        ">"
```
Comment (dweindl) [inline swig/edata.i:40]: I think including some info on whether preequilibration or presimulation is required could be helpful.

**Review by dweindl — APPROVED** [review]

File: swig/edata.i [hunk 4]
```diff
@@ -8,5 +8,51 @@ using namespace amici;
...
+    ])
+%}
+%extend amici::ExpData {
+%pythoncode %{
+def __repr__(self):
```
Comment (dweindl) [inline swig/edata.i:45]: Could also be set for ExpDataView. iirc, the default repr isn't very helpful there either.

Comment (FFroehlich) [reply swig/edata.i:45]: 👍  done for swigptrview

File: swig/edata.i (hunk 2 above)
Comment (FFroehlich) [reply swig/edata.i:25]: 👍

File: python/sdist/amici/numpy.py [hunk 5]
```diff
@@ -141,6 +141,14 @@ def __deepcopy__(self, memo):
         other._cache = copy.deepcopy(self._cache)
         return other
 
+    def __repr__(self):
+        """
+        String representation of the object
+
+        :returns: string representation
+        """
+        return f'{self.__class__.__name__}({self._swigptr})'
```
Comment (dweindl) [inline python/sdist/amici/numpy.py:150]:
```suggestion
        return f'<{self.__class__.__name__}({self._swigptr})>'
```

**Review by dweindl — APPROVED** [review]
:+1:

**Pull request merged**